helix-core/src/diff.rs

```diff
@@ -0,0 +1,72 @@
+use ropey::Rope;
+
+use crate::{Change, Transaction};
+
+/// Compares `old` and `new` to generate a [`Transaction`] describing
+/// the steps required to get from `old` to `new`.
+pub fn compare_ropes(old: &Rope, new: &Rope) -> Transaction {
+    use std::borrow::Cow;
+
+    // `similar` only works on contiguous data, so a `Rope` has
+    // to be temporarily converted into a vec until the diff
+    // is created.
+    let old_vec: Vec<_> = old.lines().map(Cow::from).collect();
+    let new_vec: Vec<_> = new.lines().map(Cow::from).collect();
+
+    // A timeout is set so after 5 seconds, the algorithm will start
+    // approximating. This is especially important for big `Rope`s or
+    // `Rope`s that are extremely dissimilar so the diff will be
+    // created in a reasonable amount of time.
+    //
+    // Note: Ignore the clippy warning, as the trait bounds of
+    // `Transaction::change()` require an iterator implementing
+    // `ExactIterator`.
+    let time = std::time::Instant::now() + std::time::Duration::from_secs(5);
```

Comment:
I assume Helix blocks the UI on this, right?  If that's the case, I would make this timeout way shorter.  The max I would consider acceptable is 1 second, and ideally I'd prefer it to be 0.2 or 0.3 seconds.

It won't get as good diffs for huge files that way, but that's a perfectly fine tradeoff I think, until we get Helix's architecture to be non-blocking on this kind of stuff.